### PR TITLE
#399 Extension module cancel workflow

### PR DESCRIPTION
**Pull request by JudeNiroshan**
In this PR, I added a cancel endpoint to the Adyen integration.

TODO:
- [x] manual test

### CODE REVIEW

**Review by lojzatran — APPROVED** [review]

File: extension/docs/CancelPayment.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+**Please see [Integration Guide](WebComponentsIntegrationGuide.md) first before continuing with this document.**
+
+## Cancel a payment
+If you have authorised a payment but do not want to capture it, you need to make a Cancel request to release the funds back to the shopper.
+If the payment has already been captured, you need to [refund it](./Refund.md) instead.
```
Comment (JudeNiroshan) [inline extension/docs/CancelPayment.md:5]:
```suggestion
If you have authorised payment but do not want to capture it, then you can make a cancel request to release the funds back to the shopper.
If the payment has already been captured, then you should [refund it](./Refund.md) instead.
```

File: extension/docs/CancelPayment.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    "type": "CancelAuthorization",
+    "amount": {
+      "currencyCode": "EUR",
+      "centAmount": 500
+    },
+    "state": "Initial"
+  }
+}
+```
+Extension module will update the transaction with `Pending` transaction state and `interactionId` field with value `pspReference` from the Adyen response:
```
Comment (JudeNiroshan) [inline extension/docs/CancelPayment.md:29]:
```suggestion
Extension module will update the CTP transaction with `Pending` transaction state and `interactionId` field with value `pspReference` from the Adyen response:
```

File: extension/docs/WebComponentsIntegrationGuide.md [hunk 3]
```diff
@@ -612,5 +613,13 @@ By default, payments are captured immediately (or with [delay](https://docs.adye
 
 If you need to explicitly request a capture for each payment please follow our [manual capture documentation](./ManualCapture.md).
 
+# Cancel or refund
+If you want to return the funds to your shopper, use either Cancel or Refund functionalities.
+
+This will either:
+
+- [**Cancel**](CancelPayment.md) - cancel the authorisation on an uncaptured payment.
```
Comment (JudeNiroshan) [inline extension/docs/WebComponentsIntegrationGuide.md:621]:
```suggestion
- [**Cancel**](CancelPayment.md) - cancel the authorisation on an uncaptured payment(full payment).
```

File: extension/src/paymentHandler/payment-handler.js [hunk 4]
```diff
@@ -69,4 +77,9 @@ function _isKlarna (makePaymentRequestObj) {
     && PAYMENT_METHOD_TYPE_KLARNA_METHODS.includes(makePaymentRequestObj.paymentMethod.type)
 }
 
+function _isCancelPayment (paymentObject) {
+  return getAuthorizationTransactionSuccess(paymentObject) &&
+    getCancelAuthorizationTransactionInit(paymentObject)
```
Comment (JudeNiroshan) [inline extension/src/paymentHandler/payment-handler.js:82]: If we have previously tried to cancel this same payment, will we have any remaining Init Transactions?

File: extension/src/service/web-component-service.js [hunk 5]
```diff
@@ -24,6 +24,10 @@ function manualCapture (manualCaptureRequestObj) {
   return callAdyen(`${config.adyen.legacyApiBaseUrl}/capture`, manualCaptureRequestObj)
 }
 
+function cancelPayment (cancelRequestObj) {
+  return callAdyen(`${config.adyen.legacyApiBaseUrl}/cancel`, cancelRequestObj)
```
Comment (JudeNiroshan) [inline extension/src/service/web-component-service.js:28]: if we have no internet access and failed this call?

File: extension/docs/CancelPayment.md (hunk 1 above)
Comment (lojzatran) [reply extension/docs/CancelPayment.md:5]: This is copied from Adyen 😊 I guess they have it professionally written, so I won't make your fixes. https://docs.adyen.com/checkout/cancel (second paragraph).

File: extension/src/paymentHandler/payment-handler.js (hunk 4 above)
Comment (lojzatran) [reply extension/src/paymentHandler/payment-handler.js:82]: If you tried, it should update the status to Pending.

File: extension/src/service/web-component-service.js (hunk 5 above)
Comment (lojzatran) [reply extension/src/service/web-component-service.js:28]: Then it will not be cancelled and the whole update will fail. The shop has to repeat.

**Pull request merged**